README.md
bandwidthd-pSQL-frontend
========================

PHP frontend to display bandwidthd data that was stored in a postgreSQL database

The origin of this project is: [http://sourceforge.net/projects/bandwidthd](http://sourceforge.net/projects/bandwidthd)
All glory goes to the author of [bandwidthd](http://sourceforge.net/projects/bandwidthd)

The main aim of this partly fork is to make the PHP files more robust against SQL injections.
I'm starting this fork because the author of the original [bandwidthd does no development on it anymore](http://sourceforge.net/p/bandwidthd/discussion/308608/thread/cef76694/#7b3f) 
but I'm very happy to merge it back if someone has access to the sourceforge project page.

See [CHANGELOG](https://github.com/individual-it/bandwidthd-pSQL-frontend/blob/master/CHANGELOG.md#changelog) for full details about the changes.

DATABASE SUPPORT of bandwidthd
------------------------------

Since version 2.0, Bandwidthd now has support for external databases.  This system 
consists of 3 major parts:

1. The Bandwidthd binary which acts as a sensor, recording traffic information and 
storing it in a database across the network or on the local host.  In this mode 
Bandwidthd uses very little ram and CPU. In addition, multiple sensors can record 
to the same database.

2. The database system.  Currently Bandwidthd only supports Postgresql.

3. The webserver and php application. Bundled with Bandwidthd in the "phphtdocs" 
directory is a php application that reports on and graphs the contents of the database.  
This has been designed to be easy to customize.  Everything is passed around on the urls, 
just tinker with it a little and you'll see how to generate custom graphs pretty easy.
**The files you find in this repository replace the files mentioned above.**

Using Bandwidthd with a database has many advantages, such as much lower overhead, because 
graphs are only graphed on demand.  And much more flexibility, SQL makes building new 
reports easy, and php+sql greatly improves the interactivity of the reports.

My ISP has now switched over to the database driven version of bandwidthd entirely, we 
have half a dozen sensors sprinkled around the country, writing millions of data points a 
day on our customers into the system.

## INSTRUCTIONS

As a prerequisite for these instructions, you must have Postgresql installed and working, 
as well as a web server that supports php.

### Database Setup:
1. Create a database for Bandwidthd.  You will need to create users that can access the 
database remotely if you want remote sensors.

2. Bandwidthd's schema is in "schema.postgresql".  "psql mydb username < schema.postgresql" 
should load it and create the 2 tables and 4 indexes.Thi

### Bandwidthd Setup:
1. Add the following lines to your bandwidthd.conf file:

        # Standard postgres connect string, just like php, see postgres docs for details
        pgsql_connect_string "user = someuser dbname = mydb password = secret host = databaseserver.com"
        # Arbitrary sensor name, I recommend the sensors fully qualified domain name
        sensor_id "sensor1.mycompany.com"
        # Tells Bandwidthd to keep no data and preform no graphing locally
        graph false
        # If this is set to true Bandwidthd will try to recover the daily log 
        # into the database.  If you set this true on purpose only do it once.
        # Bandwidthd does not track the fact that it has already transferred 
        # certain records into the database.
        recover_cdf false

2. Simply start bandwidthd, and after a few minutes data should start appearing in 
your database.  If not, check syslog for error messages.


### Web Server Setup:
1. Copy the contents of this repository into your web tree some where.
2. Edit config.php to set your db connect string

You should now be able to access the web application and see you graphs.  All graphing 
is done by graph.php,  all parameters are passed to it in it's url.  You can create 
custom urls to pull custom graphs from your own index pages, or use the canned 
reporting system.

In addition, you should schedule bd_pgsql_purge.sh to run every so often.  I recomend 
running it weekly.  This script outputs sql statements that aggregate the older 
data points in your database in order to reduce the amount of data that needs to
be slogged through in order to generate yearly, monthly, and weekly graphs.

Example:

    bd_pgsql_purge.sh | psql bandwidthd postgres

Will connect to the bandwidthd database on local host as the user postgres and summarize 
the data.



CHANGELOG.md
CHANGELOG
---------
What changed compared to the original project
* 04.09.2013 - made the website HTML validated, changed css to a newer more modern style (Bootstrap v3)
* 13.08.2013 - sanitizing all $_GET inputs
* 15.08.2013 - protect the config file from beeing read directly
* 16.08.2012 - add table sorting

details.php
<?include("include.php");?>
<!DOCTYPE HTML>
<html>
<head>
	<title>BandWidthd</title>
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1.0">
	<script src="http://code.jquery.com/jquery-1.10.1.min.js"></script>
	<script type="text/javascript" src="js/jquery.tinysort.min.js"></script>
	<script type="text/javascript" src="js/bandwidthd.js"></script>
	<link media="screen" rel="stylesheet" type="text/css" href="css/style.css">
	<link media="screen" rel="stylesheet" type="text/css" href="css/bootstrap.min.css">
</head>
<body>
	<div class="container content">
<img alt="logo" src="logo.gif">
<? trim_get ();

if (isset($_GET['sensor_id'])) {
	$sensor_id = filter_var($_GET['sensor_id'], FILTER_SANITIZE_NUMBER_INT);
}
else
{
	echo "<br>Please provide a sensor_id";
    exit(1);
}

if (isset($_GET['ip'])) {
	$ip =  sanitize_ip ($_GET['ip']);
}
else
{
    echo "<br>Please provide an ip address";
    exit(1);
}
                                                                                                                             
echo "<h3>";
if (strpos($ip, "/") === FALSE) echo "$ip - ".gethostbyaddr($ip)."</h3>";
else echo "Total - $ip</h3>";

$db = ConnectDb();

if ($ip == "0.0.0.0/0")
{
    $rxtable = "bd_rx_total_log";
	$txtable = "bd_tx_total_log";
}
else
{
    $rxtable = "bd_rx_log";
	$txtable = "bd_tx_log";
}

$sql = "select rx.scale as rxscale, tx.scale as txscale, tx.total+rx.total as total, tx.total as sent,
rx.total as received, tx.tcp+rx.tcp as tcp, tx.udp+rx.udp as udp,
tx.icmp+rx.icmp as icmp, tx.http+rx.http as http,
tx.p2p+rx.p2p as p2p, tx.ftp+rx.ftp as ftp
from
                                                                                                                             
(SELECT ip, max(total/sample_duration)*8 as scale, sum(total) as total, sum(tcp) as tcp, sum(udp) as udp, sum(icmp) as icmp,
sum(http) as http, sum(p2p) as p2p, sum(ftp) as ftp
from sensors, $txtable
where sensors.sensor_id = '$sensor_id'
and sensors.sensor_id = ".$txtable.".sensor_id
and ip <<= '$ip'
group by ip) as tx,
                                                                                                                             
(SELECT ip, max(total/sample_duration)*8 as scale, sum(total) as total, sum(tcp) as tcp, sum(udp) as udp, sum(icmp) as icmp,
sum(http) as http, sum(p2p) as p2p, sum(ftp) as ftp
from sensors, $rxtable
where sensors.sensor_id = '$sensor_id'
and sensors.sensor_id = ".$rxtable.".sensor_id
and ip <<= '$ip'
group by ip) as rx
                                                                                                                             
where tx.ip = rx.ip;";
//echo "</center><pre>$sql</pre><center>";exit(0);
$result = pg_query($sql);
echo "<div class='table-responsive'><table class='table table-hover table-bordered' id='xtable' style='font-size:13px;'>
		<thead>
			<tr>
				<th title='click for sorting' onclick='sortTable(0);'>IP</th>
				<th title='click for sorting' onclick='sortTable(1);'>Name</th>
				<th title='click for sorting' onclick='sortTable(2);'>Total</th>
				<th title='click for sorting' onclick='sortTable(3);'>Sent</th>
				<th title='click for sorting' onclick='sortTable(4);'>Received</th>
				<th title='click for sorting' onclick='sortTable(5);'>tcp</th>
				<th title='click for sorting' onclick='sortTable(6);'>udp</th>
				<th title='click for sorting' onclick='sortTable(7);'>icmp</th>
				<th title='click for sorting' onclick='sortTable(8);'>http</th>
				<th title='click for sorting' onclick='sortTable(9);'>p2p</th>
				<th title='click for sorting' onclick='sortTable(10);'>ftp</th>
			</tr>
		</thead><tbody><tr><td>";
$r = pg_fetch_array($result);

if (strpos($ip, "/") === FALSE) echo "$ip</td><td>".gethostbyaddr($ip) . '</td>';
else echo "Total<td>$ip</td>";

echo fmtb($r['total']).fmtb($r['sent']).fmtb($r['received']).
	fmtb($r['tcp']).fmtb($r['udp']).fmtb($r['icmp']).fmtb($r['http']).
    fmtb($r['p2p']).fmtb($r['ftp']);
echo "</tbody></table>";


	echo "<div class='panel panel-default'><div class='panel-heading'><h3 class='panel-title'>Daily</h3></div>
	    <div class='panel-body'>
	    	<div class='well'>
		    	Send:<br><img alt='graph' src='graph.php?ip=$ip&amp;sensor_id=".$sensor_id."&amp;table=$txtable&amp;yscale=".(max($r['txscale'], $r['rxscale']))."'><br>
				<img alt='' src=legend.gif><br>
			</div>
			<div class='well'>
				Receive:<br><img alt='graph' src='graph.php?ip=$ip&amp;sensor_id=".$sensor_id."&amp;table=$rxtable&amp;yscale=".(max($r['txscale'], $r['rxscale']))."'>
				<img alt='' src=legend.gif><br>
			</div>
		</div>
	</div>";

	echo "<div class='panel panel-default'><div class='panel-heading'><h3 class='panel-title'>Weekly</h3></div>
	    <div class='panel-body'>
	    	<div class='well'>
		    	Send:<br><img alt='graph' src='graph.php?interval=".INT_WEEKLY."&amp;ip=$ip&amp;sensor_id=$sensor_id&amp;table=$txtable&amp;yscale=".(max($r['txscale'], $r['rxscale']))."'><br>
				<img alt='' src=legend.gif><br>
			</div>
			<div class='well'>
				Receive:<br><img alt='graph' src='graph.php?interval=".INT_WEEKLY."&amp;ip=$ip&amp;sensor_id=$sensor_id&amp;table=$rxtable&amp;yscale=".(max($r['txscale'], $r['rxscale']))."'>
				<img alt='' src=legend.gif><br>
			</div>
		</div>
	</div>";

	echo "<div class='panel panel-default'><div class='panel-heading'><h3 class='panel-title'>Monthly</h3></div>
	    <div class='panel-body'>
	    	<div class='well'>
		    	Send:<br><img alt='graph' src='graph.php?interval=".INT_MONTHLY."&amp;ip=$ip&amp;sensor_id=$sensor_id&amp;table=$txtable&amp;yscale=".(max($r['txscale'], $r['rxscale']))."'><br>
				<img alt='' src=legend.gif><br>
			</div>
			<div class='well'>
				Receive:<br><img alt='graph' src='graph.php?interval=".INT_MONTHLY."&amp;ip=$ip&amp;sensor_id=$sensor_id&amp;table=$rxtable&amp;yscale=".(max($r['txscale'], $r['rxscale']))."'>
				<img alt='' src=legend.gif><br>
			</div>
		</div>
	</div>";

	echo "<div class='panel panel-default'><div class='panel-heading'><h3 class='panel-title'>Yearly</h3></div>
	    <div class='panel-body'>
	    	<div class='well'>
		    	Send:<br><img alt='graph' src='graph.php?interval=".INT_YEARLY."&amp;ip=$ip&amp;sensor_id=$sensor_id&amp;table=$txtable&amp;yscale=".(max($r['txscale'], $r['rxscale']))."'><br>
				<img alt='' src=legend.gif><br>
			</div>
			<div class='well'>
				Receive:<br><img alt='graph' src='graph.php?interval=".INT_YEARLY."&amp;ip=$ip&amp;sensor_id=$sensor_id&amp;table=$rxtable&amp;yscale=".(max($r['txscale'], $r['rxscale']))."'>
				<img alt='' src=legend.gif><br>
			</div>
		</div>
	</div>

</div>";

include('footer.php');

include.php
<?
define("INT_DAILY", 60*60*24*2);
define("INT_WEEKLY", 60*60*24*8);
define("INT_MONTHLY", 60*60*24*35);
define("INT_YEARLY", 60*60*24*400);

define("XOFFSET", 90);
define("YOFFSET", 45);

$config_conf_key = 1;

require("config.php");

/**
 * trims all $_GET inputs
 */
function trim_get () 
{
	//trim all inputs
	foreach ($_GET as $key => $value) {
		$_GET[$key] = trim($_GET[$key]);
	}
	
}

/**
 * 
 * @param string (IP) $given_ip
 * @return string (sanitized IP)
 */
function sanitize_ip ($given_ip) 
{
	if (preg_match("/(1?[1-9]?[0-9]|2?(?:[0-4]?[0-9]|5[0-5]))\.(1?[1-9]?[0-9]|2?(?:[0-4]?[0-9]|5[0-5]))\.(1?[1-9]?[0-9]|2?(?:[0-4]?[0-9]|5[0-5]))\.(1?[1-9]?[0-9]|2?(?:[0-4]?[0-9]|5[0-5]))(\/[0-9]{1,2})?\b/", $given_ip,$ip))
	{
		return $ip[0];
	}
	else
	{
		return "0.0.0.0/0";
	}
}

function ConnectDb()
{
	global $db_connect_string;

    $db = pg_pconnect($db_connect_string);
    if (!$db)
    {
        printf("DB Error, could not connect to database");
        exit(1);
     }
    return($db);
}
                                                                                                                             
function fmtb($kbytes)
{
	$Max = 1024;
	$Output = $kbytes;
	$Suffix = 'K';

	if ($Output > $Max)
	{
		$Output /= 1024;
		$Suffix = 'M';
	}

	if ($Output > $Max)
	{
		$Output /= 1024;
		$Suffix = 'G';
	}

	if ($Output > $Max)
	{
		$Output /= 1024;
		$Suffix = 'T';
	}
		//return(sprintf("<td sort='%d' align=right><tt>%d</td>",$kbytes, $kbytes));
		
	return(sprintf("<td class='%d' style='text-align:right;'>%.1f%s</td>",$kbytes, $Output, $Suffix));
}

//TODO check for needed functions and extensions. e.g. 	imagecreate / PHP5_GD

$starttime = time();
set_time_limit(300);
?>

graph.php
<?
require("include.php");

// Returns x location of any given timestamp
function ts2x($ts)
	{
	global $timestamp, $width, $interval;
	return(($ts-$timestamp)*(($width-XOFFSET) / $interval) + XOFFSET);
	}

// If we have multiple IP's in a result set we need to total the average of each IP's samples
function AverageAndAccumulate()
	{
	global $Count, $total, $icmp, $udp, $tcp, $ftp, $http, $p2p, $YMax;
	global $a_total, $a_icmp, $a_udp, $a_tcp, $a_ftp, $a_http, $a_p2p;
	
	foreach ($Count as $key => $number)
    	{
	    $total[$key] /= $number;
    	$icmp[$key] /= $number;
    	$udp[$key] /= $number;
    	$tcp[$key] /= $number;
    	$ftp[$key] /= $number;
    	$http[$key] /= $number;
    	$p2p[$key] /= $number;
    	}

	foreach ($Count as $key => $number) 
	{
		if (!isset($a_total[$key])) {
			$a_total[$key] = 0;
		}
		if (!isset($a_icmp[$key])) {
			$a_icmp[$key] = 0;
		}
		if (!isset($a_udp[$key])) {
			$a_udp[$key] = 0;
		}
		if (!isset($a_tcp[$key])) {
			$a_tcp[$key] = 0;
		}
		if (!isset($a_ftp[$key])) {
			$a_ftp[$key] = 0;
		}
		if (!isset($a_http[$key])) {
			$a_http[$key] = 0;
		}
		if (!isset($a_p2p[$key])) {
			$a_p2p[$key] = 0;
		}
			
			
			
		$a_total[$key] += $total[$key];
		$a_icmp[$key] += $icmp[$key];
		$a_udp[$key] += $udp[$key];
		$a_tcp[$key] += $tcp[$key];
		$a_ftp[$key] += $ftp[$key];
		$a_http[$key] += $http[$key];
		$a_p2p[$key] += $p2p[$key];

		if ($a_total[$key] > $YMax)
			$YMax = $a_total[$key];
		}
	
	unset($GLOBALS['total'], $GLOBALS['icmp'], $GLOBALS['udp'], $GLOBALS['tcp'], $GLOBALS['ftp'], $GLOBALS['http'], $GLOBALS['p2p'], $GLOBALS['Count']);

	$total = array();
	$icmp = array();
	$udp = array();
	$tcp = array();
	$ftp = array();
	$http = array();
	$p2p = array();
	$Count = array();
	}
 

$db = ConnectDb();

// Get parameters
trim_get ();

if (isset($_GET['width']))
    $width = filter_var($_GET['width'], FILTER_SANITIZE_NUMBER_INT); 
else
	$width = DFLT_WIDTH;

if (isset($_GET['height']))
    $height = filter_var($_GET['height'], FILTER_SANITIZE_NUMBER_INT);  
else
	$height = DFLT_HEIGHT;

if (isset($_GET['interval']))
    $interval = filter_var($_GET['interval'], FILTER_SANITIZE_NUMBER_INT); 
else
	$interval = DFLT_INTERVAL;

if (isset($_GET['ip']))
    $ip =  sanitize_ip ($_GET['ip']);
else
	exit(1);

if (isset($_GET['sensor_id']))
	$sensor_id = filter_var($_GET['sensor_id'], FILTER_SANITIZE_NUMBER_INT);  
else
	exit(1);

if (isset($_GET['timestamp']))
    $timestamp = filter_var($_GET['timestamp'], FILTER_SANITIZE_NUMBER_INT); 
else
	$timestamp = time() - $interval + (0.05*$interval);

//make sure just real tables can be accessed
if (isset($_GET['table']) 
	&& ($_GET['table'] == "bd_tx_total_log" || $_GET['table'] == "bd_tx_log" || $_GET['table'] == "bd_rx_total_log" || $_GET['table'] == "bd_rx_log"))
    $table = $_GET['table'];
else
	$table = "bd_rx_log";

if (isset($_GET['yscale']))
    $yscale = filter_var($_GET['yscale'], FILTER_SANITIZE_NUMBER_INT); 

$total = array();
$icmp = array();
$udp = array();
$tcp = array();
$ftp = array();
$http = array();
$p2p = array();
$Count = array();

// Accumulator
$a_total = array();
$a_icmp = array();
$a_udp = array();
$a_tcp = array();
$a_ftp = array();
$a_http = array();
$a_p2p = array();

$sql = "select *, extract(epoch from timestamp) as ts from sensors, $table where sensors.sensor_id = ".$table.".sensor_id and ip <<= '$ip' and sensors.sensor_id = '$sensor_id' and timestamp > $timestamp::abstime and timestamp < ".($timestamp+$interval)."::abstime order by ip;";
//echo $sql."<br>"; exit(1);
$result = pg_query($sql);

// The SQL statement pulls the data out of the database ordered by IP address, that way we can average each
// datapoint for each IP address to provide smoothing and then toss the smoothed value into the accumulator
// to provide accurate total traffic rate.

$last_ip = "";
$SentPeak = 0;
$TotalSent = 0;
while ($row = pg_fetch_array($result))
	{
	if ($row['ip'] != $last_ip)
		{
		AverageAndAccumulate();
		$last_ip = $row['ip'];
		}

	$x = ($row['ts']-$timestamp)*(($width-XOFFSET)/$interval)+XOFFSET;
	$xint = (int) $x;

	//echo "xint: ".$xint."<br>";
	if (!isset($Count[$xint])) {
		$Count[$xint] = 0;
	}
	$Count[$xint]++;

	if (!isset($total[$xint])) {
		$total[$xint] = 0;
	}
	if (!isset($icmp[$xint])) {
		$icmp[$xint] = 0;
	}
	if (!isset($udp[$xint])) {
		$udp[$xint] = 0;
	}
	if (!isset($tcp[$xint])) {
		$tcp[$xint] = 0;
	}		
	if (!isset($ftp[$xint])) {
		$ftp[$xint] = 0;
	}
	if (!isset($http[$xint])) {
		$http[$xint] = 0;
	}
	if (!isset($p2p[$xint])) {
		$p2p[$xint] = 0;
	}		
	
	
	if ($row['total']/$row['sample_duration'] > $SentPeak)
		$SentPeak = $row['total']/$row['sample_duration'];
	$TotalSent += $row['total'];
	$total[$xint] += $row['total']/$row['sample_duration'];
	$icmp[$xint] += $row['icmp']/$row['sample_duration'];
	$udp[$xint] += $row['udp']/$row['sample_duration'];
	$tcp[$xint] += $row['tcp']/$row['sample_duration'];
	$ftp[$xint] += $row['ftp']/$row['sample_duration'];
	$http[$xint] += $row['http']/$row['sample_duration'];
	$p2p[$xint] += $row['p2p']/$row['sample_duration'];                                                                                                                             
	}

// One more time for the last IP
AverageAndAccumulate();

// Pull the data out of Accumulator
$total = $a_total;
$icmp = $a_icmp;
$udp = $a_udp;
$tcp = $a_tcp;
$ftp = $a_ftp;
$http = $a_http;
$p2p = $a_p2p;

$YMax += $YMax*0.05;    // Add an extra 5%

// if a y scale was specified override YMax
if (isset($yscale))
    $YMax = $yscale/8;

// Plot the data

header("Content-type: image/png");

$im = imagecreate($width, $height);
$white = imagecolorallocate($im, 255, 255, 255);
$yellow = ImageColorAllocate($im, 255, 255, 0);
$purple = ImageColorAllocate($im, 255, 0, 255);
$green  = ImageColorAllocate($im, 0, 255, 0);
$blue   = ImageColorAllocate($im, 0, 0, 255);
$lblue  = ImageColorAllocate($im, 128, 128, 255);
$brown  = ImageColorAllocate($im, 128, 0, 0);
$red    = ImageColorAllocate($im, 255, 0, 0);
$black  = ImageColorAllocate($im, 0, 0, 0);

for($Counter=XOFFSET+1; $Counter < $width; $Counter++)
	{
	if (isset($total[$Counter]))
		{

		//otherwise we might run into Division by zero problems
		if ($YMax != 0) {
			
			// Convert the bytes/sec to y coords
	        $total[$Counter] = ($total[$Counter]*($height-YOFFSET))/$YMax;
			$tcp[$Counter] = ($tcp[$Counter]*($height-YOFFSET))/$YMax;
	        $ftp[$Counter] = ($ftp[$Counter]*($height-YOFFSET))/$YMax;
			$http[$Counter] = ($http[$Counter]*($height-YOFFSET))/$YMax;
			$p2p[$Counter] = ($p2p[$Counter]*($height-YOFFSET))/$YMax;
	        $udp[$Counter] = ($udp[$Counter]*($height-YOFFSET))/$YMax;
			$icmp[$Counter] = ($icmp[$Counter]*($height-YOFFSET))/$YMax;
		} else {
			$total[$Counter] = 0;
			$tcp[$Counter] = 0;
			$ftp[$Counter] = 0;
			$http[$Counter] = 0;
			$p2p[$Counter] = 0;
			$udp[$Counter] = 0;
			$icmp[$Counter] = 0;
		}
			
		// Stack 'em up!
		// Total is stacked from the bottom
		// Icmp is on the bottom too
		// Udp is stacked on top of icmp
		$udp[$Counter] += $icmp[$Counter];
		// TCP and p2p are stacked on top of Udp
		$tcp[$Counter] += $udp[$Counter];
		$p2p[$Counter] += $udp[$Counter];
 		// Http is stacked on top of p2p
		$http[$Counter] += $p2p[$Counter];
		// Ftp is stacked on top of http
        $ftp[$Counter] += $http[$Counter];

		// Plot them!
		//echo "$Counter:".$Counter." (h-y)-t:".($height-YOFFSET) - $total[$Counter]." h-YO-1:".$height-YOFFSET-1;
        ImageLine($im, $Counter, ($height-YOFFSET) - $total[$Counter], $Counter, $height-YOFFSET-1, $yellow);
        ImageLine($im, $Counter, ($height-YOFFSET) - $icmp[$Counter], $Counter, $height-YOFFSET-1, $red);
        ImageLine($im, $Counter, ($height-YOFFSET) - $udp[$Counter], $Counter, ($height-YOFFSET) - $icmp[$Counter] - 1, $brown);
        ImageLine($im, $Counter, ($height-YOFFSET) - $tcp[$Counter], $Counter, ($height-YOFFSET) - $udp[$Counter] - 1, $green);
        ImageLine($im, $Counter, ($height-YOFFSET) - $p2p[$Counter], $Counter, ($height-YOFFSET) - $udp[$Counter] - 1, $purple);
        ImageLine($im, $Counter, ($height-YOFFSET) - $http[$Counter], $Counter, ($height-YOFFSET) - $p2p[$Counter] - 1, $blue);
        ImageLine($im, $Counter, ($height-YOFFSET) - $ftp[$Counter], $Counter, ($height-YOFFSET) - $http[$Counter] - 1, $lblue);
		}
//	else
//		echo $Counter." not set<br>";
	}                                                                                                                             

// Margin Text
if ($table == "bd_tx_total_log" || $table == "bd_tx_log")
	$tx_rx_text = "Send";
else 
	$tx_rx_text = "Transfered";

	//|| $_GET['table'] == "bd_rx_total_log" || $_GET['table'] == "bd_rx_log"))
	
	
	
if ($SentPeak < 1024/8)
	$txtPeakSendRate = sprintf("Peak %s Rate: %.1f KBits/sec", $tx_rx_text, $SentPeak*8);
else if ($SentPeak < (1024*1024)/8)
    $txtPeakSendRate = sprintf("Peak %s Rate: %.1f MBits/sec", $tx_rx_text, ($SentPeak*8.0)/1024.0);
else 
	$txtPeakSendRate = sprintf("Peak %s Rate: %.1f GBits/sec", $tx_rx_text, ($SentPeak*8.0)/(1024.0*1024.0));
                                                                                                                             
if ($TotalSent < 1024)
	$txtTotalSent = sprintf("%s %.1f KBytes", $tx_rx_text, $TotalSent);
else if ($TotalSent < 1024*1024)
	$txtTotalSent = sprintf("%s %.1f MBytes", $tx_rx_text, $TotalSent/1024.0);
else 
	$txtTotalSent = sprintf("%s %.1f GBytes", $tx_rx_text, $TotalSent/(1024.0*1024.0));
                                                                                                                             
ImageString($im, 2, XOFFSET+5,  $height-20, $txtTotalSent, $black);
ImageString($im, 2, $width/2+XOFFSET/2,  $height-20, $txtPeakSendRate, $black);

// Draw X Axis

ImageLine($im, 0, $height-YOFFSET, $width, $height-YOFFSET, $black);

// Day/Month Seperator bars

if ((24*60*60*($width-XOFFSET))/$interval > ($width-XOFFSET)/10)
	{
	$ts = getdate($timestamp);
	$MarkTime = mktime(0, 0, 0, $ts['mon'], $ts['mday'], $ts['year']);
	
    $x = ts2x($MarkTime);
    while ($x < XOFFSET)
    	{
        $MarkTime += (24*60*60);
	    $x = ts2x($MarkTime);
        }
                                                                                                                             
    while ($x < ($width-10))
    	{
        // Day Lines
        ImageLine($im, $x, 0, $x, $height-YOFFSET, $black);
        ImageLine($im, $x+1, 0, $x+1, $height-YOFFSET, $black);
                                                                                                                             
        $txtDate = strftime("%a, %b %d", $MarkTime);
        ImageString($im, 2, $x-30,  $height-YOFFSET+10, $txtDate, $black);
                                                                                                                             
        // Calculate Next x
        $MarkTime += (24*60*60);
	    $x = ts2x($MarkTime);
        }
	}
else if ((24*60*60*30*($width-XOFFSET))/$interval > ($width-XOFFSET)/10)
	{
	// Monthly Bars
	$ts = getdate($timestamp);
	$month = $ts['mon'];
	$MarkTime = mktime(0, 0, 0, $month, 1, $ts['year']);
	
    $x = ts2x($MarkTime);
    while ($x < XOFFSET)
    	{
		$month++;
        $MarkTime = mktime(0, 0, 0, $month, 1, $ts['year']);
	    $x = ts2x($MarkTime);
        }
                                                                                                                             
    while ($x < ($width-10))
    	{
        // Day Lines
        ImageLine($im, $x, 0, $x, $height-YOFFSET, $black);
        ImageLine($im, $x+1, 0, $x+1, $height-YOFFSET, $black);
                                                                                                                             
        $txtDate = strftime("%b, %Y", $MarkTime);
        ImageString($im, 2, $x-25,  $height-YOFFSET+10, $txtDate, $black);
                                                                                                                             
        // Calculate Next x
		$month++;
        $MarkTime = mktime(0, 0, 0, $month, 1, $ts['year']);
	    $x = ts2x($MarkTime);
        }
	}
else
	{
	// Year Bars
    $ts = getdate($timestamp);
    $year = $ts['year'];
    $MarkTime = mktime(0, 0, 0, 1, 1, $year);
                                                                                                                             
    $x = ts2x($MarkTime);
    while ($x < XOFFSET)
        {
        $year++;
        $MarkTime = mktime(0, 0, 0, 1, 1, $year);
        $x = ts2x($MarkTime);
        }
                                                                                                                             
    while ($x < ($width-10))
        {
        // Day Lines
        ImageLine($im, $x, 0, $x, $height-YOFFSET, $black);
        ImageLine($im, $x+1, 0, $x+1, $height-YOFFSET, $black);
                                                                                                                             
        $txtDate = strftime("%b, %Y", $MarkTime);
        ImageString($im, 2, $x-25,  $height-YOFFSET+10, $txtDate, $black);
                                                                                                                             
        // Calculate Next x
        $year++;
        $MarkTime = mktime(0, 0, 0, 1, 1, $year);
        $x = ts2x($MarkTime);
        }	
	}

// Draw Major Tick Marks
if ((6*60*60*($width-XOFFSET))/$interval > 10) // pixels per 6 hours is more than 2
	$MarkTimeStep = 6*60*60; // Major ticks are 6 hours
else if ((24*60*60*($width-XOFFSET))/$interval > 10)
	$MarkTimeStep = 24*60*60; // Major ticks are 24 hours;
else if ((24*60*60*30*($width-XOFFSET))/$interval > 10)
	{
	// Major tick marks are months
	$MarkTimeStep = 0; // Skip the standard way of drawing major tick marks below

    $ts = getdate($timestamp);
    $month = $ts['mon'];
    $MarkTime = mktime(0, 0, 0, $month, 1, $ts['year']);
                                                                                                                             
    $x = ts2x($MarkTime);
    while ($x < XOFFSET)
        {
        $month++;
        $MarkTime = mktime(0, 0, 0, $month, 1, $ts['year']);
        $x = ts2x($MarkTime);
        }
                                                                                                                             
    while ($x < ($width-10))
        {
        // Day Lines
		$date = getdate($MarkTime);
		if ($date['mon'] != 1)
			{
	        ImageLine($im, $x, $height-YOFFSET-5, $x, $height-YOFFSET+5, $black);                                                                                                                      
    	    $txtDate = strftime("%b", $MarkTime);
        	ImageString($im, 2, $x-5,  $height-YOFFSET+10, $txtDate, $black);
          	}
                                                                                                                   
        // Calculate Next x
        $month++;
        $MarkTime = mktime(0, 0, 0, $month, 1, $ts['year']);
        $x = ts2x($MarkTime);
        }
	}
else
	$MarkTimeStep = 0; // Skip Major Tick Marks

if ($MarkTimeStep)
	{
	$ts = getdate($timestamp);
	$MarkTime = mktime(0, 0, 0, $ts['mon'], $ts['mday'], $ts['year']);
	$x = ts2x($MarkTime);

	while ($x < ($width-10))
		{
    	if ($x > XOFFSET) 
			{
    	    ImageLine($im, $x, $height-YOFFSET-5, $x, $height-YOFFSET+5, $black);
	        }
		$MarkTime += $MarkTimeStep;
	    $x = ts2x($MarkTime);
		}
	}

// Draw Minor Tick marks
if ((60*60*($width-XOFFSET))/$interval > 4) // pixels per hour is more than 2
	$MarkTimeStep = 60*60;  // Minor ticks are 1 hour
else if ((6*60*60*($width-XOFFSET))/$interval > 4)
	$MarkTimeStep = 6*60*60; // Minor ticks are 6 hours
else if ((24*60*60*($width-XOFFSET))/$interval > 4)
	$MarkTimeStep = 24*60*60;
else
	$MarkTimeStep = 0; // Skip minor tick marks

if ($MarkTimeStep)
	{
	$ts = getdate($timestamp);
	$MarkTime = mktime(0, 0, 0, $ts['mon'], $ts['mday'], $ts['year']);
	$x = ts2x($MarkTime);

	while ($x < ($width-10))
		{
    	if ($x > XOFFSET) 
			{
    	    ImageLine($im, $x, $height-YOFFSET, $x, $height-YOFFSET+5, $black);
	        }
		$MarkTime += $MarkTimeStep;
	    $x = ts2x($MarkTime);
		}
	}

// Draw Y Axis
ImageLine($im, XOFFSET, 0, XOFFSET, $height, $black);

$YLegend = 'k';
$Divisor = 1;
if ($YMax*8 > 1024*2)
	{
    $Divisor = 1024;    // Display in m
    $YLegend = 'm';
    }

if ($YMax*8 > 1024*1024*2)
	{
    $Divisor = 1024*1024; // Display in g
    $YLegend = 'g';
	}

if ($YMax*8 > 1024*1024*1024*2)
	{
    $Divisor = 1024*1024*1024; // Display in t
    $YLegend = 't';
    }
                                                                                                                             
$YStep = $YMax/10;
if ($YStep < 1)
	$YStep=1;
$YTic=$YStep;
                                                                                                                             
while ($YTic <= ($YMax - $YMax/10))
	{
    $y = ($height-YOFFSET)-(($YTic*($height-YOFFSET))/$YMax);
	ImageLine($im, XOFFSET, $y, $width, $y, $black);
    $txtYLegend = sprintf("%4.1f %sbits/s", (8.0*$YTic)/$Divisor, $YLegend);
    ImageString($im, 2, 3, $y-7, $txtYLegend, $black);
	$YTic += $YStep;
	}

imagepng($im); 
imagedestroy($im);